payara-management-api/src/test/java/fish/payara/arquillian/management/mockimpl/ClientExtension.java

```diff
@@ -0,0 +1,101 @@
+ *    recipient has the option to distribute your version of this file under
+ *    either the CDDL, the GPL Version 2 or to extend the choice of license to
+ *    its licensees as provided above.  However, if you add GPL Version 2 code
+ *    and therefore, elected the GPL Version 2 license, then the option applies
+ *    only if the new code is made subject to such option by the copyright
+ *    holder.
+ */
+
+package fish.payara.arquillian.management.mockimpl;
+
+import fish.payara.arquillian.management.ExternalServerManagement;
+import org.jboss.arquillian.container.spi.event.container.AfterStart;
+import org.jboss.arquillian.container.test.spi.client.deployment.AuxiliaryArchiveAppender;
+import org.jboss.arquillian.core.api.InstanceProducer;
+import org.jboss.arquillian.core.api.annotation.ApplicationScoped;
+import org.jboss.arquillian.core.api.annotation.Inject;
+import org.jboss.arquillian.core.api.annotation.Observes;
+import org.jboss.arquillian.core.spi.InvocationException;
+import org.jboss.arquillian.core.spi.LoadableExtension;
+
+/**
+ * Example of how container implementations should interact with management-api.
+ * A container should register a producers
+ */
+public class ClientExtension implements LoadableExtension {
+    @Override
+    public void register(ExtensionBuilder extensionBuilder) {
+        // this registers client-side management interface
+        extensionBuilder.observer(ExternalManagementProducer.class);
+
+        // If we'd utilize remote protocol, we'd use these
+        //extensionBuilder.service(AuxiliaryArchiveAppender.class, ClientArchiveAppender.class);
+
+        // but our mock container is embedded one, and therefore we perform client registration right here
+        ContainerExtension quasiContainerExtension = new ContainerExtension();
+        quasiContainerExtension.register(extensionBuilder);
+
+        // debug arquillian events
+        extensionBuilder.observer(Observer.class);
+    }
+
+    static class Observer {
+        @SuppressWarnings("unused")
+        public void whatsUp(@Observes Object o) {
+            System.out.printf("%s: %s\n", o.getClass(), o);
+            if (o instanceof Throwable) {
+                // rethrow, otherwise all tests pass
+                throw new InvocationException((Throwable)o);
+            }
+        }
+
+    }
+
+
+    /**
+     * External management is applied in Arquillian Client Context
+     */
+    static class ExternalManagementProducer {
+        @Inject @ApplicationScoped //application scope is only one active when class rule is invoked
+        InstanceProducer<ExternalServerManagement> externalServerManagementInstanceProducer;
+
+        // likely the best point to inject stuff is after container is started, as it is properly configured
+        // another good point would be AfterSetup, or observing container itself fired as event
+        @SuppressWarnings("unused")
+        public void produce(@Observes AfterStart event) {
```

Comment:
This is the core of implementation side -- producing an instance at correct time.